### PR TITLE
Added ntp_server

### PR DESCRIPTION
**Pull request by jonnytdevops**
Hey Everyone,

Here is my initial code for ntp_server for the cisco node util gem. I know I'll need to write a test before this can be merged.

Comments are very welcome.

Cheers

JT (Jonny)
Puppet Labs

### CODE REVIEW

File: lib/cisco_node_utils.rb [hunk 1]
```diff
@@ -26,6 +26,7 @@
 require 'cisco_node_utils/snmpgroup'
 require 'cisco_node_utils/snmpserver'
 require 'cisco_node_utils/snmpuser'
+require 'cisco_node_utils/ntp_server'
```
Comment (glennmatthews) [inline lib/cisco_node_utils.rb]: move this up - we're trying to keep these alphabetized.

File: lib/cisco_node_utils/command_reference_common.yaml [hunk 2]
```diff
@@ -452,6 +452,18 @@ inventory:
     config_get: "show inventory"
     config_get_token: ["TABLE_inv", "ROW_inv", 0, "vendorid"]
 
+ntp_server:
+  server:
+    config_get: "show running-config all | include 'ntp server'"
+    config_get_token: '/(?:[0-9]{1,3}\.){3}[0-9]{1,3}/'
```
Comment (glennmatthews) [inline lib/cisco_node_utils/command_reference_common.yaml]: I'd prefer to see this regexp anchored a bit more - perhaps something like:

`config_get_token: '/ntp server ((?:[0-9]{1,3}\.){3}[0-9]{1,3})/'`

File: lib/cisco_node_utils/ntp_server.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    attr_reader :name, :prefer
+
+    def initialize(ntpserver_id, prefer, instantiate=true)
+      @ntpserver_id = ntpserver_id.to_s
+      @ntpserver_prefer = prefer
+      fail ArgumentError,
+           'Invalid value(IP is not an IP address)' unless @ntpserver_id[/^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/]
+      fail ArgumentError,
+           'Invalid value(prefer must be true or false)' unless @ntpserver_prefer == :true ||
+                                                                @ntpserver_prefer == :false ||
```
Comment (glennmatthews) [inline lib/cisco_node_utils/ntp_server.rb]: I'd rather use `true`/`false` unless there's a good reason to use symbols instead.

**Comment by jonnytdevops** [comment]
Hi @glennmatthews 
I've updated the code as per your comments. Feedback appreciated.

We use :true and :false in our Puppet code, however as per your comments, I'm now using boolean true and false in the cisco util gem and doing the conversion inside the puppet provider.

Thanks

JT

File: tests/test_ntp_server.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    assert_includes(Cisco::NtpServer.ntpservers, id2)
+    assert_equal(Cisco::NtpServer.ntpservers[id1], ntp1)
+    assert_equal(Cisco::NtpServer.ntpservers[id2], ntp2)
+
+    ntp1.destroy
+    ntp2.destroy
+    refute_includes(Cisco::NtpServer.ntpservers, id1)
+    refute_includes(Cisco::NtpServer.ntpservers, id2)
+  end
+end
```
Comment (glennmatthews) [inline tests/test_ntp_server.rb]: This is causing a Rubocop failure:

```
tests/test_ntp_server.rb:75:4: C: Final newline missing.
end
```

File: lib/cisco_node_utils/ntp_server.rb [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      ntpservers_list.each do |id|
+        hash[id] = NtpServer.new(id, preferred_servers.include?(id), false)
+      end
+
+      hash
+    end
+
+    def create
+      config_set('ntp_server', 'server', state: '', ip: @ntpserver_id,
+                  prefer: @ntpserver_prefer == true ? 'prefer' : '')
```
Comment (glennmatthews) [inline lib/cisco_node_utils/ntp_server.rb]: Not sure the == true is needed:

`prefer: @ntpserver_prefer ? 'prefer' : '')`

**Comment by glennmatthews** [comment]
One minor nitpick and one fix needed to satisfy Rubocop, but otherwise this looks great. Don't forget to update CHANGELOG.md too :-)

:+1:

**Comment by jonnytdevops** [comment]
@glennmatthews thank for your feedback. I've made the requested change, cheers!

**Comment by chrisvanheuveln** [comment]
:+1:  ship it

**Pull request merged**